L1Trigger/CSCTriggerPrimitives/src/CSCComparatorDigiFitter.cc

```diff
@@ -0,0 +1,290 @@
+    float z_tmp = 0.0;
+
+    // ignore layers with no digis
+    if (p.second.size()==0) continue;
+
+    // loop on all matching digis in this layer
+    for (const auto& hit: p.second) {
+      const float fractional_strip = hit.getFractionalStrip();
+      const auto& layer_geo = cscChamber->layer(detId.layer())->geometry();
+      const float wire = layer_geo->middleWireOfGroup(stub.getKeyWG() + 1);
+
+      // get the phi of each comparator digi
+      const LocalPoint& csc_intersect = layer_geo->intersectionOfStripAndWire(fractional_strip, wire);
+      const GlobalPoint& csc_gp = cscGeometry_->idToDet(detId)->surface().toGlobal(csc_intersect);
+      const float gpphi = csc_gp.phi();
+
+      // normalize phi values according to first one
+      if (phis_.size()>0 and gpphi>0 and phis_[0]<0 and  (gpphi-phis_[0])>M_PI)
+        phi_tmp += (gpphi-2*M_PI);
+      else if (phis_.size()>0 and gpphi<0 and phis_[0]>0 and (gpphi-phis_[0])<-M_PI)
+        phi_tmp += (gpphi+2*M_PI);
+      else
+        phi_tmp += (csc_gp.phi());
+
+      z_tmp = csc_gp.z();
+      radius_tmp += csc_gp.perp();
+    }
+
+    //in case there are more than one comparator digis in one layer
+    radius_tmp = radius_tmp/(p.second).size();
+    radius_ += radius_tmp;
+
+    zs_.push_back(z_tmp);
+    ezs_.push_back(0);
+
+    phi_tmp = phi_tmp/(p.second).size();
+    phis_.push_back(phi_tmp);
+    ephis_.push_back(cscHalfStripWidth(detId)/sqrt(12));
+  }
+}
+
+void CSCComparatorDigiFitter::fit(const CSCDetId& ch_id, const CSCCorrelatedLCTDigi& stub, const CSCComparatorDigiCollection& hCSCComparators)
+{
+  // first, match the comparator digis to the LCT
+  matchingComparatorDigisLCT(ch_id, stub, hCSCComparators);
+
+  // second, get the coordinates
+  getComparatorDigiCoordinates(ch_id, stub);
+
+  // get radius of the stub from key layer
+  const CSCDetId key_id(ch_id.endcap(), ch_id.station(), ch_id.ring(), ch_id.chamber(), CSCConstants::KEY_CLCT_LAYER);
+  const float fractional_strip = stub.getFractionalStrip();
+
+  const auto& cscChamber = cscGeometry_->chamber(ch_id);
+  const auto& layer_geo = cscChamber->layer(CSCConstants::KEY_CLCT_LAYER)->geometry();
+
+  // LCT::getKeyWG() also starts from 0
+  const float wire = layer_geo->middleWireOfGroup(stub.getKeyWG() + 1);
+  const LocalPoint& csc_intersect = layer_geo->intersectionOfStripAndWire(fractional_strip, wire);
+  const GlobalPoint& csc_gp = cscGeometry_->idToDet(key_id)->surface().toGlobal(csc_intersect);
+
+  // get radius from key layer
+  if (useKeyRadius_)
+    radius_ = radius_/phis_.size();
+  else
+    radius_ = csc_gp.perp();
+  
+  float alpha = -99., beta = 0.;
+  // do a fit to the comparator digis
+  calculateSlopeIntercept(alpha, beta);
+  if (phis_ .size() <= 2 or std::abs(alpha)>=99){
+    alpha = csc_gp.phi();
+    beta = 0.0;
+  }
+
+  float stripPhiPitch = layer_geo->stripPhiPitch();
+  if (nStripBits_)
+    stripPhiPitch = stripPhiPitch/nStripBits_;
+
+  int fit_z_layers[6];
```

Comment:
"6" has been replaced by CSCConstants::NUM_LAYERS, an enum which has the value 6.